### PR TITLE
Rebase ocean/dev on master

### PR DESCRIPTION
**Pull request by glwagner**
### Description

This PR rebases `ocean/dev` onto `master`. In doing so it supercedes #1723, which was intended to merge improvements to `Ocean` into `ocean/dev`, and whose CI was failing due to CI pipeline improvements on master that were not incorporated on `ocean/dev`.

It also adds a new example that simulates the Eady instability problem, and contains a number of improvements to `SpectralElementField` and `domain` that were originally intended for a follow-up PR to #1723.

Much this is essentially WIP, yet it is imperative to merge into master in its current state so that we can continue with forward progress on `Ocean`.

During the rebase, some of the `Ocean` tests may have been broken and will need to be fixed in the course of tidying up this PR for merger.

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by glwagner** [comment]
bors try

**Comment by glwagner** [comment]
bors try

**Comment by glwagner** [comment]
bors try

**Comment by glwagner** [comment]
bors try

**Comment by glwagner** [comment]
bors try

File: tutorials/Ocean/eady_instability.jl [hunk 1]
```diff
@@ -0,0 +1,331 @@
+# # Eady instability
+#
+# This script simulates the instability of a sheared, free-surface
+# flow using `ClimateMachine.Ocean.HydrostaticBoussinesqSuperModel`.
+
+using Printf
+
+ENV["GKSwstype"] = "nul"
```
Comment (charleskawczynski) [inline tutorials/Ocean/eady_instability.jl:8]:
```suggestion
```
I think this is already set when the docs are run

Comment (glwagner) [reply tutorials/Ocean/eady_instability.jl:8]: True --- but it's not set when we run examples locally. I can still remove it but then it has to be added in to manually run examples on headless systems, which is annoying. Please advise.

**Comment by glwagner** [comment]
bors try

**Comment by glwagner** [comment]
bors try

**Comment by glwagner** [comment]
bors try

**Comment by glwagner** [comment]
@christophernhill @jm-c some integration tests are failing, but it looks like only on the GPU? Can you take a look? Here's the GPU simple box test:

https://buildkite.com/clima/climatemachine-ci/builds/827#8fb8a64f-7fa8-4db9-9edb-164decdbf15c

**Comment by glwagner** [comment]
bors try

**Comment by jm-c** [comment]
Hi Greg, will take a look at this afternoon. Thanks. Jean-Michel

File: src/Ocean/SplitExplicit01/SplitExplicitLSRK3nMethod.jl [hunk 2]
```diff
@@ -264,7 +264,7 @@ function dostep!(
         iter_tot = 0
         for i in 1:nImplSteps
             solve_time = @elapsed iters =
-                linearsolve!(lm!, ivdc_solver, ivdc_Q, ivdc_RHS)
+                linearsolve!(lm!, nothing, ivdc_solver, ivdc_Q, ivdc_RHS)
```
Comment (christophernhill) [inline src/Ocean/SplitExplicit01/SplitExplicitLSRK3nMethod.jl:267]: @jm-c did you mean to change this line. The ``nothing`` doesn't look right to me.

Comment (christophernhill) [reply src/Ocean/SplitExplicit01/SplitExplicitLSRK3nMethod.jl:267]: @jm-c I see now that there was an external change to introduce a pre-conditioner, so this makes sense for that. 

Not really sure why multiple-dispatch wasn't used for this change.

Comment (jm-c) [reply src/Ocean/SplitExplicit01/SplitExplicitLSRK3nMethod.jl:267]: @christophernhill This addition was there in the LSRK2nMethod (but I did not track when it was added). I made similar addition to the LSRK3nMethod so that it would run. Since the state-check are passing for simple_box_rk3.jl and simple_dbl_gyre.jl I assume it's fine.

**Comment by jm-c** [comment]
bors try

**Comment by jm-c** [comment]
bors try

**Comment by glwagner** [comment]
Squashing commits:

```
4c4209cc7 Add restart feature and fix initial vtk output
6a3b93b43 New Split-Explicit scheme using LowStorageRungeKutta3N
298aee626 Minor changes that fix the restart
b9f30ec57 Simple-Box problem using SplitExplicitLSRK3nSolver
d48a60aa3 add (debug) print, but commented out for now
6c69ec670 add one more print (to check)
475a08949 Fix missing reset of dR
956b3e6b8 minor edit + skip some print
2bda9e374 Only do reconciliation on Eta at lastStage
b243d3253 Simplify
f05e92386 Update refvals of SplitExplicitLSRK2nSolver tests
1dd654a5c Clean-up simple-box test for new split-explicit RK-3 scheme
a65b7d5a9 Only proc 0 remove output dir
0afd537f9 Update simple_dbl_gyre.jl example
b5d841a68 Fix FluxSecondOrder BC in SplitExplicit01 code
73c4acddf Increase horizontal viscosity in simple double-gyre test
dcf237a49 Reorganizes OceanProblems and add InitialValueProblem
9c05d4d5a Adds a test for InitialValueProblem plus formatting update
b116cd4da Adds CartesianDomains + tutorial for geostrophic adjustment
858db3364 Small style updates to geostrophic adjustment problem;
998db5905 Updates + radiating eddy tutorial
c62729b78 Completes docstring for InitialValueProblem
b58f75965 Glue not join
ab458ee91 Update cartesian_field.jl
2ff9a7970 Better explanation of boundary conditions
825965e10 Faster gravity waves...
8b55a4eac CartesianField needs only the MPIStateArray
05013368b Implements SuperHydrostaticBoussinesqModel and internal wave example
035a8e6f2 Moar abstraction
9a74253eb Format base files in CLiMA style  o fixed base files in ocean/dev needed climaformat.jl to pass CI
01f29fc18 Adds CartesianDomains and two new tutorials
282d10592 Updates
e9fc59bfe Less fragile input checking for assemble, and finally good internal wave parameters
a15ee79c0 Just call it solver_configuration in SuperModel
02a04042f Fixes a few typos
e8572dc49 New way to specify experiments + many new ocean tutorials
0e325ecb8 Shuffles imports and turbulence -> turbulence_closure
f4c3495a7 Adds CartesianDomains and two new tutorials
1b8e6ea4d Makes domains and fields more rational
ba516956d Simple non-breaking functions-as-stresses interface
b6da40b63 Fixes infinite loop in specifying Stress and adds double gyre example
c3b403030 Double gyre tutorial
d72ba1edd Use global setting array_type as default in domain constructor
5d0a81852 Differentiate barotropic and baroclinic double gyre
a4e5554b9 Applies formatter
edb84f310 Slims down assemble implementation
867bce376 Fixes a few bugs in rectangular elements and fields
593e73cbe Eady problem write up in the example
9952b9d21 Updates
a6acbe6da More views and better constructor for RectangularElement
938bed924 Adds forced bickley script
90b5bb785 Updates tutorials
d6c648975 Adds a JLD2Writer with changes to SpectralElementField for checkpointing
929dec376 Minor bugfixes for JLD2Writer
23c968309 Implements a generic user-defined forcing term
b1061588e Dont splat tuple in source!
a7e2cb645 Begins implementation of a state filter
07e3251ea Copy-paste from ocean/glw/eady-instability
eb52f0851 Puts tutorials in proper Ocean directory;
0a059be9a Remove barotropic gyre and forced bickley example
949c33ae3 Grinding through the rebase
683ae15c2 Irons out some rebase wrinkles
6a1f622e9 Puts the shear back in shear_instability
09c37c536 Adds missing imports in internal_wave tutorial
767a41545 Reduces time-step for baroclinic double gyre example
5a2b0b837 Parameter adjustment for eady instability example
e48dd60db Applies formatter
e2a87aaa3 Fix up tests using RectangularDomain
645f77d2a Tweaks to fix tests, plus removes redundant AbstractOceanProblem definition
12ab7225a Access coords through local_geometry
2fe1e0ff1 Apply formatter
fe2151f66 More apply formatter
d44658915 Adds some ocean-y docstrings to the docs/src/APIs
541705bbb Omit docstrings of InitialValueProblem and InitialCondition since they arent exported
13926d4f2 JLD2Writers is in Ocean for now
dd341ad2f Updates list of tutorials
800a53c6f Dont use Revise in tutorials of course
c79b1558f Fix literate issue for shear_instability
b631eee35 Fix another literate issue for shear_instability
df752ecc7 Desperate hacks to get Literate to work
1f97955a2 Local_geometry foo
7d7d4c94a Extracts SplitExplicit driver config from Drivers.jl
01054b878 Applies formatter
764364032 Fixes irregularity in construction of OceanModel in simple box ivd test
73d531834 revert some changes from PR #1554
ba046360d fix previous commit
fab90c11b Fix boundary-condition settings
76107a540 simplify BC (fix run on GPU)
```

**Comment by glwagner** [comment]
bors try

**Comment by glwagner** [comment]
bors try

**Review by christophernhill — APPROVED** [review]

**Review by jm-c — APPROVED** [review]
Looks good to me.

**Comment by glwagner** [comment]
bors try

**Comment by christophernhill** [comment]
bors try

**Comment by christophernhill** [comment]
timed out - again....

bors r+
